src/VisualStudio/Core/Next/Remote/JsonRpcSession.cs

```diff
@@ -31,15 +31,35 @@ internal class JsonRpcSession : RemoteHostClient.Session
         private readonly CancellationTokenRegistration _cancellationRegistration;
 
         public static async Task<JsonRpcSession> CreateAsync(
-            PinnedRemotableDataScope snapshot,
+            Func<CancellationToken, Task<PinnedRemotableDataScope>> getSnapshotAsync,
             object callbackTarget,
             Stream serviceStream,
             Stream snapshotStreamOpt,
             CancellationToken cancellationToken)
         {
-            var session = new JsonRpcSession(snapshot, callbackTarget, serviceStream, snapshotStreamOpt, cancellationToken);
+            var snapshot = getSnapshotAsync == null ? null : await getSnapshotAsync(cancellationToken).ConfigureAwait(false);
+
+            JsonRpcSession session;
+            try
+            {
+                session = new JsonRpcSession(snapshot, callbackTarget, serviceStream, snapshotStreamOpt, cancellationToken);
+            }
+            catch
+            {
+                snapshot?.Dispose();
+                throw;
+            }
 
-            await session.InitializeAsync().ConfigureAwait(false);
+            try
+            {
+                await session.InitializeAsync().ConfigureAwait(false);
+            }
+            catch when (!cancellationToken.IsCancellationRequested)
+            {
+                // The session disposes of itself when cancellation is requested.
+                session.Dispose();
```

Comment:
can I ask what it is throwing? StreamJson disposed exception? do you know where those are being thrown? they should be translated to cancellation exception. otherwise, it is a bug.